### ISSUE TITLE
Is it possible to limit the number of predictions a Yolov8 object detector makes?

### ISSUE THREAD

**Issue by iaverypadberg** [ultralytics/ultralytics#3876]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Quick background- Yolov8 object detectors converted to tflite have an output tensor of candidate detections with dimensions (batches, 4 + num_classes, num_candidate_detections). Each candidate detection is made up of (xc,yy,w,h, class_1_conf, class_2_conf,...,class_N_conf) as described in [this]([https://discuss.roboflow.com/t/interpreting-yolov8-tflite-output/1562/2]) post. 

For my 3 class detector my output tensor has a shape of (1,7,2100). I usually squeeze this array and then transpose it so it then has a shape of (2100,7). That is 2100 suggested predictions that I then need to sort through and filter out low confidence detections. Since I'm looking to deploy this model on an edge computing device, iterating over this array becomes costly and balloons the time it takes to post process(~+600ms).

Other object detectors, such as tensorflows [SpaghettiNet model ]([https://discuss.roboflow.com/t/interpreting-yolov8-tflite-output/1562/2]) allow you to specify some post processing parameters in the training pipeline.config, one of which is the max_total_detections.
```
post_processing {
      batch_non_max_suppression {
        score_threshold: 1e-8
        iou_threshold: 0.6
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SIGMOID
    }
```
As a result of including these flags during training the number of predictions is limited to 100. This saves a significant amount of time when filtering the models' output tensor. 

Is there something similar for Yolov8 models? Perhaps a training flag or a or something in the export of the model from Pytorch -> tflite?

**Comment by glenn-jocher** [comment]
@iaverypadberg hello! Appreciate the detailed question. 

Yes, it is possible to limit the number of predictions in YOLOv8. This can be achieved by tweaking the post-processing stage of our detection pipeline known as Non-Maximum Suppression (NMS). 

In YOLOv8, NMS controls the number of output predictions and essentially filters out low confidence or overlapping detections. There are two parameters that can be modified - 'Confidence threshold' and 'IOU threshold'. 

'Confidence threshold' determines the minimum confidence score a detection must have to be reported. If you increase this threshold, you will get fewer detections. 'IOU threshold' is used to determine whether two boxes overlap too much, such that they likely correspond to the same object. By increasing this threshold, you are likely to remove duplicate detections and hence get fewer predictions.

However, please note that this change doesn't occur during training but during prediction or evaluation stages, and these parameters are located in our test.py script. 

While we do not currently support limiting the detections directly in the training process, we continue to work on feature additions and improvements based on user feedback.

As for the export to tflite, currently, there is no direct support to limit the output detection thresholds, but with some custom script, you could add a few post-processing steps similar to what you've seen with SpaghettiNet model.

I hope this explanation helps. If you have any other questions, feel free to ask!

**Comment by quietbamboo** [comment]
Actually, with increased iou value, we get more predictions then smaller iou. So it looks like smaller iou value, we allow detections with less overlap, so we get less detections?

**Comment by glenn-jocher** [comment]
@quietbamboo certainly! Allow me to clarify the operation of IOU threshold in the YOLOv8 Non-Maximum Suppression (NMS) process.

The Intersection Over Union (IOU) threshold is used to determine whether two bounding boxes overlap significantly or not. Specifically, if the IOU of two boxes exceeds the IOU threshold, they are considered to overlap too much, and the one with a lower confidence score is suppressed.

To your point, a lower IOU threshold means bounding boxes need to have less overlap to be considered for the same object, resulting in fewer final detected objects. On the other hand, if we increase the IOU threshold, bounding boxes require more overlap to be considered the same object, leading to more detections. One might get more detections with a higher IOU threshold because it becomes more permissible for multiple bounding boxes to exist for the same object, hence leading to more detections.

However, it's important to remember, this is a balance - too low of an IOU threshold could lead to missing valid detections, while a high threshold might result in multiple detections for the same object. Determining the optimal value often involves some experimentation and depends on factors like the nature of the dataset and specific requirements of your use case.

I hope this helps clarify the purpose and effect of the IOU threshold in the YOLOv8 detector. If you have any further questions, feel free to ask!

**Comment by quietbamboo** [comment]
@glenn-jocher Much appreciated for the detailed answer. A followup question:

In one of our applications, we wish to detect digits, however, some detections are returned with bounding boxes of 0 width, which is obviously wrong (however, the location of the detection is also incorrect to my surprise), but the detected class type is correct somehow. This phenomenon is not common though. <img width="592" alt="x" src="https://github.com/ultralytics/ultralytics/assets/836556/2d6a2fac-10d2-4dc1-9a87-bd4f13d57ee4">

I guess this is because one correct detection is suppressed by this wrong detection due to conf issue. So can we limit bounding box size (or maybe in the future)? Or the only way I can do this is to give IOU=1.0 and write my customized NMS?

**Comment by glenn-jocher** [comment]
@quietbamboo, thank you for your follow-up question.

The issue with some detections returning zero-width bounding boxes, despite having the correct class identification, is likely due to an anomaly in the model's prediction rather than a suppression issue in NMS. The bounding box regression part of the model can sometimes provide incorrect dimensions for various reasons, such as training data noise, model overfitting, or convergence problems.

As for bounding box size constraints, there isn't a direct built-in feature in YOLOv8 to limit detection by minimum or maximum bounding box size during the post-processing steps. However, it is a reasonable feature consideration for quality control in certain applications, and we may consider such additions for future versions.

Until such features are potentially introduced, a custom post-processing script could be used to filter out detections with unrealistic dimensions (e.g., width or height being zero or below an empirically determined threshold).

Using an IOU threshold of 1.0 is not recommended, as it would assume perfect overlap for suppression, effectively disabling the NMS process. With IOU=1.0, you would retain all detections, including duplicates for the same object, which would lead to a high false-positive rate.

Writing customized NMS that considers bounding box size as an additional criterion can be a good workaround for now. In this custom NMS, you can eliminate detections with bounding boxes that don't meet certain dimensional criteria before applying the traditional confidence and IOU based filtering.

In summary, consider creating a post-processing script to filter out unrealistic bounding boxes until size filtering can be integrated into the model's prediction pipeline or as a feature in future releases of YOLOv8.

Wishing you the best with your digit detection application! If you have further questions or concerns, please feel free to reach out.

**Comment by quietbamboo** [comment]
@glenn-jocher Thank you for your detailed response. Yes, I'm looking forward to see the YOLOv8 update with bounding box size limit parameter.

Just to add to the above example, the model actually is correct. The 0-size bounding box issue is not really due to training data noise, model overfitting, or convergence problems. Because when I run with torch.device('cpu'), the predicted bounding box is perfectly fine, however, when I switch to torch.device('mps') on my Macbook Pro for obvious speedup reason, 0-size bounding box begins to appears. So it looks like its a mps compatibility issue which you guys might fix in the future.

**Comment by glenn-jocher** [comment]
@quietbamboo, thanks for bringing this specific observation to our attention. The discrepancy in bounding box sizes when switching between 'cpu' and 'mps' devices indeed suggests that the issue you're encountering may be related to Metal Performance Shaders (MPS) compatibility rather than the model itself.

Using 'mps', which is specific to Apple's Metal API, can introduce certain variances due to differences in how computations are performed on the GPU versus the CPU. Some of these differences could arise from the precision of operations, optimization paths, or even potential bugs in the MPS implementation.

At Ultralytics, we strive to ensure compatibility across different platforms and hardware. Issues like these are critical for us to know as they may affect users who are deploying models to edge devices with specific acceleration frameworks. We regularly update our libraries and models to maintain robustness and compatibility.

Please do keep an eye on our GitHub repository and the documentation at https://docs.ultralytics.com for updates on YOLOv8, including any improvements regarding device compatibility issues such as the one you've described with MPS.

In the meantime, as a workaround, you may continue to perform inference on the 'cpu' to avoid this issue or implement an additional check in your post-processing to filter out any detection with a bounding box size of zero when using 'mps'.

We appreciate your understanding and patience as we work to address these complex challenges. If you notice anything else or have further questions, your insights and inquiries are always welcome!
